taxcalc/parameters.py

```diff
@@ -1,170 +1,496 @@
-                           " values for the years between {0} and {1}.")
-                    raise ValueError(msg.format(self.current_year, year))
-
-            # Set up the '_X = [a, b,...]' variables as 'X = a'
-            self.set_year(self._current_year)
+        if self.current_year != self.start_year:
+            self.set_year(self.start_year)
+        if not reform:
+            return  # no reform to implement
+        reform_years = sorted(list(reform.keys()))
+        last_reform_year = max(reform_years)
+        if last_reform_year > self.end_year:
+            msg = 'reform provision in year={} > end_year={}'
+            ValueError(msg.format(last_reform_year, self.end_year))
+        for year in reform_years:
+            if year != self.start_year:
+                self.set_year(year)
+            self._update({year: reform[year]})
+        self.set_year(self.start_year)
 
     @property
     def current_year(self):
+        """
+        Current policy parameter year property.
+        """
         return self._current_year
 
     @property
     def start_year(self):
+        """
+        First policy parameter year property.
+        """
         return self._start_year
 
     @property
-    def budget_years(self):
-        return self._budget_years
+    def num_years(self):
+        """
+        Number of policy parameter years property.
+        """
+        return self._num_years
 
-    def increment_year(self):
-        self._current_year += 1
-        self.set_year(self._current_year)
+    @property
+    def end_year(self):
+        """
+        Last policy parameter year property.
+        """
+        return self._end_year
 
-    def set_year(self, yr):
-        for name, vals in self._vals.items():
+    def set_year(self, year):
+        """
+        Set policy parameters to values for specified calendar year.
+
+        Parameters
+        ----------
+        year: int
+            calendar year for which to current_year and parameter values
+
+        Raises
+        ------
+        ValueError:
+            if year is not in [start_year, end_year] range.
+
+        Returns
+        -------
+        nothing: void
+
+        Notes
+        -----
+        To increment the current year, use the following statement::
+
+            ppo.set_year(ppo.current_year + 1)
+
+        where ppo is a policy Parameters object.
+        """
+        if year < self.start_year or year > self.end_year:
+            msg = 'year passed to set_year() must be in [{},{}] range.'
+            raise ValueError(msg.format(self.start_year, self.end_year))
+        self._current_year = year
+        year_zero_indexed = year - self._start_year
+        for name in self._vals:
             arr = getattr(self, name)
-            setattr(self, name[1:], arr[yr - self._start_year])
+            setattr(self, name[1:], arr[year_zero_indexed])
+
+    @classmethod
+    def default_data(cls, metadata=False, start_year=None):
```

Comment:
Can you explain what you mean by "the problems you are having"? In this thread, I'm attempting to explain why this branch won't run the functionality in dropq. Is that what you mean by "the problems you are having?" Here are some ways to reproduce the issue:

- run `test_make_Calculator_increment_years_first` as it exists in master on this branch, changing only the `budget_years` keyword arg from master to `num_years`. Then run pytest as follows:

```
py.test taxcalc/tests/test_calculate.py -k test_make_Calculator_increment_years_first
```

A second way to reproduce is as I mentioned above:

- install via `python setup.py install` taxcalc synced with this branch, checkout the latest master of dropq, then run `example.py` at the top of the repo.

Then you will see the assertion violation. Are you saying that you cannot reproduce this behavior? 

```
Traceback (most recent call last):
  File "example.py", line 14, in <module>
    mY_dec, mX_dec, df_dec, mY_bin, mX_bin, df_bin, fiscal_tots = dropq.run_models(tax_dta, num_years=10, user_mods=user_mods)
  File "/Users/talumbau/dev/dropq/dropQ/dropq/dropq.py", line 320, in run_models
    calc3 = calculator(params=params3, records=records3, mods=user_mods)
  File "/Users/talumbau/miniconda3/envs/heroku_aei/lib/python2.7/site-packages/taxcalc-0.2_644_g13a470e-py2.7.egg/taxcalc/calculate.py", line 63, in calculator
  File "/Users/talumbau/miniconda3/envs/heroku_aei/lib/python2.7/site-packages/taxcalc-0.2_644_g13a470e-py2.7.egg/taxcalc/calculate.py", line 95, in __init__
AssertionError
```